File: week1_baseline/rust/03_prompt_builder/src/message.rs
use std::fmt;

pub struct Message {
    pub role: String,
    pub content: String,
    pub tool_use_id: Option<String>,
}

impl Message {
    pub fn new(role: impl Into<String>, content: impl Into<String>, tool_use_id: Option<String>) -> Self {
        Self { role: role.into(), content: content.into(), tool_use_id }
    }
}

impl fmt::Display for Message {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let id_tag = self
            .tool_use_id
            .as_deref()
            .map(|id| format!(" [{id}]"))
            .unwrap_or_default();
        let content: String = self.content.chars().take(61).collect();
        write!(f, "#<Message role={}{} content={}...>", self.role, id_tag, content)
    }
}

File: week1_baseline/python/07_the_run_dsl/lib/boukensha/message.py
from dataclasses import dataclass


@dataclass
class Message:
    role: str
    content: str
    tool_use_id: str = None

    def __repr__(self):
        id_tag = f" [{self.tool_use_id}]" if self.tool_use_id else ""
        return f"#<Message role={self.role}{id_tag} content={self.content[:61]}...>"

    __str__ = __repr__
